core/src/main/java/io/grpc/internal/GrpcUtil.java

```diff
@@ -743,6 +752,22 @@ public InternalLogId getLogId() {
     return null;
   }
 
+  /** Gets stream tracers based on CallOptions. */
+  static ClientStreamTracer[] getClientStreamTracers(
+      CallOptions callOptions, boolean isTransparentRetry) {
+    List<ClientStreamTracer.Factory> factories = callOptions.getStreamTracerFactories();
+    ClientStreamTracer[] tracers = new ClientStreamTracer[factories.size() + 1];
+    StreamInfo streamInfo = StreamInfo.newBuilder()
+        .setCallOptions(callOptions)
+        .setIsTransparentRetry(isTransparentRetry)
+        .build();
+    for (int i = 0; i < factories.size(); i++) {
+      tracers[i] = factories.get(i).newClientStreamTracer(streamInfo);
+    }
+    tracers[tracers.length - 1] = NOOP_TRACER;
```

Comment:
Add a comment about why this is present.

Reply:
Done.